va_args implementation for AAPCS.

JamieCunliffe opened the pull request:
Implement the va args in codegen for AAPCS, this will be used as the
default va_args implementation for AArch64 rather than the va_args
llvm-ir as it currently is.

This should fix the following issues:
https://github.com/rust-lang/rust/issues/56475
https://github.com/rust-lang/rust/issues/72579

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @petrochenkov (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

petrochenkov commented:
r? @nikic (I guess)

bjorn3 commented on src/librustc_codegen_llvm/builder.rs:
```diff
@@ -1062,6 +1062,15 @@ impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
+            llvm::LLVMAddIncoming(phi, vals.as_ptr(), bbs.as_ptr(), vals.len() as c_uint);
+            phi
+        }
+    }
```
Please just make the `phi` below `pub(crate)` instead. I removed `phi` from `BuilderMethods` as not every codegen backend uses phi nodes. Cranelift for example uses basic block parameters instead.

JamieCunliffe committed:
va_args implementation for AAPCS.

Implement the va args in codegen for AAPCS, this will be used as the
default va_args implementation for AArch64 rather than the va_args
llvm-ir as it currently is.

Copyright (c) 2020, Arm Limited.

JamieCunliffe commented:
@bjorn3 I addressed your comment but force pushed, sorry about that

@dlrobertson - Bringing this to your attention seeing so you commented on the issue regarding this

dlrobertson commented:
@JamieCunliffe Thanks for this! I'm not super familiar with the AAPCS64 spec, but on a cursory glance it looks great.

nikic commented:
For reference, this is the clang implementation of aapcs va_arg handling: https://github.com/llvm/llvm-project/blob/9548697df9c6f65a3dba4e8de4d015650e73101c/clang/lib/CodeGen/TargetInfo.cpp#L5795

I had a bunch of questions here, but then I realized that Rust's va_arg intrinsic only supports integer and float types, so we don't need handling for any of the "interesting" cases, like homogeneous aggregates.

My remaining question here would be why we can assume a little endian target.

nikic commented on src/librustc_codegen_llvm/va_arg.rs:
```diff
@@ -89,6 +90,82 @@ fn emit_ptr_va_arg(
+    // the offset again.
+
+    if layout.align.abi.bytes() > 8 {
+        assert!(layout.align.abi.bytes() <= 16);
```
As a mild preference, I would align to the given alignment here, even if 16 is currently the only possibility.

JamieCunliffe replied:
I think the assert was unnecessary and I was being a little over cautious, according to the AAPCS 16 is the given alignment for this (https://github.com/ARM-software/abi-aa/blob/master/aapcs64/aapcs64.rst#the-va-arg-macro).
```
if (alignof(type) > 8)
    offs = (offs + 15) & -16;   // round up
```

It should only be done on the general purpose registers though and I had missed that in the above check so I have fixed that.

dlrobertson commented:
Tested locally and things look good!

> I had a bunch of questions here, but then I realized that Rust's va_arg intrinsic only supports integer and float types, so we don't need handling for any of the "interesting" cases, like homogeneous aggregates.

Yeah IIRC we wouldn't be able to rely on the `va_arg` instruction as much (even for x86_64) if we were to support the interesting cases.

JamieCunliffe commented:
> My remaining question here would be why we can assume a little endian target.

Lack of actual hardware for testing this, I didn't have any hardware available internally to be able to run this so we decided it was best to have the assert.

JamieCunliffe committed:
Remove the assert on alignment check.

Also the alignment should only be done on general register types as
per the AAPCS so fixed that issue.

Copyright (c) 2020, Arm Limited.

nikic commented:
> > My remaining question here would be why we can assume a little endian target.
> 
> Lack of actual hardware for testing this, I didn't have any hardware available internally to be able to run this so we decided it was best to have the assert.

Okay, sounds reasonable.

@bors r+ rollup=always

bors commented:
:pushpin: Commit 8b58eb9dc49254926c91690380c065403b10fbb8 has been approved by `nikic`

<!-- @bors r=nikic 8b58eb9dc49254926c91690380c065403b10fbb8 -->
<!-- homu: {"type":"Approved","sha":"8b58eb9dc49254926c91690380c065403b10fbb8","approver":"nikic"} -->
